firmware/defmt-rtt/src/channel.rs

```diff
@@ -0,0 +1,125 @@
+use core::{
+    ptr,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+use crate::header::{SIZE, _SEGGER_RTT};
+
+/// Indicates that we are in RTT blocking-mode
+const BLOCK_IF_FULL: usize = 2;
+
+/// RTT Up channel
+#[repr(C)]
+pub(crate) struct Channel {
+    pub name: *const u8,
+    /// Pointer to the RTT buffer.
+    pub buffer: *mut u8,
+    pub size: usize,
+    /// Written by the target.
+    pub write: AtomicUsize,
+    /// Written by the host.
+    pub read: AtomicUsize,
+    pub flags: AtomicUsize,
+}
+
+impl Channel {
+    pub fn write_all(&self, mut bytes: &[u8]) {
+        // the host-connection-status is only modified after RAM initialization while the device is
+        // halted, so we only need to check it once before the write-loop
+        let write = match self.host_is_connected() {
+            true => Self::blocking_write,
+            false => Self::nonblocking_write,
+        };
+
+        while !bytes.is_empty() {
+            let consumed = write(self, bytes);
+            if consumed != 0 {
+                bytes = &bytes[consumed..];
+            }
+        }
+    }
+
+    fn blocking_write(&self, bytes: &[u8]) -> usize {
+        if bytes.is_empty() {
+            return 0;
+        }
```

Comment:
Do we need this check here? We can either move it into `write_impl` so it gets checked for both blocking and non-blocking. Or we drop it, since `write_all` won't call either write function with empty `bytes`.